src/Symfony/Component/HttpClient/MockClient.php

```diff
@@ -0,0 +1,121 @@
+use Symfony\Component\HttpClient\Chunk\ErrorChunk;
+use Symfony\Component\HttpClient\Chunk\FirstChunk;
+use Symfony\Component\HttpClient\Chunk\LastChunk;
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Component\HttpClient\Response\ResponseStream;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * Provides a way to tests the HttpClient without making actual HTTP requests.
+ *
+ * @author Gary PEGEOT <garypegeot@gmail.com>
+ */
+class MockClient implements HttpClientInterface
+{
+    /**
+     * Predefined responses. Throw a TransportExceptionInterface when none provided.
+     *
+     * @var ResponseInterface[]
+     */
+    private $responses = [];
+
+    /**
+     * MockClient constructor.
+     *
+     * @param iterable|ResponseInterface[] $responses
+     */
+    public function __construct(iterable $responses = [])
+    {
+        foreach ($responses as $response) {
+            if (!$response instanceof ResponseInterface) {
+                throw new \TypeError(sprintf('Each predefined response must an instance of %s, %s given.', ResponseInterface::class, \is_object($response) ? \get_class($response) : \gettype($response)));
+            }
+
+            $this->responses[] = $response;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if (!\count($this->responses)) {
+            throw new TransportException('No predefined response to send. Please add one or more using "addResponse" method.');
+        }
+
+        return \array_shift($this->responses);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stream($responses, float $timeout = null): ResponseStreamInterface
+    {
+        if ($responses instanceof ResponseInterface) {
+            $responses = [$responses];
+        } elseif (!\is_iterable($responses)) {
+            throw new \TypeError(sprintf('%s() expects parameter 1 to be an iterable of ResponseInterface objects, %s given.', __METHOD__, \is_object($responses) ? \get_class($responses) : \gettype($responses)));
+        }
+
+        return new ResponseStream($this->streamResponses($responses));
+    }
+
+    /**
+     * @return $this
+     */
+    public function addResponse(ResponseInterface $response)
+    {
+        $this->responses[] = $response;
+
+        return $this;
+    }
+
+    /**
+     * Clear all predefined responses.
+     */
+    public function clear(): void
+    {
+        $this->responses = [];
+    }
+
+    private function streamResponses(iterable $responses): \Generator
+    {
+        foreach ($responses as $response) {
+            $didThrow = false;
+
+            try {
+                $response->getHeaders(true);
```

Comment:
this can be move in the next try/catch and the current one can be removed

Reply:
fixed